cg/apps/demultiplex/sample_sheet/override_cycles_validator.py

```diff
@@ -0,0 +1,172 @@
+            else:
+                LOG.warning(f"Pattern {pattern} is not a valid index cycle pattern")
+                return False
+            if index_chars + ignored_chars == run_cycles and len(index_sequence) == index_chars:
+                return True
+        return False
+
+    def _validate_reads_cycles(self) -> None:
+        """
+        Determine if the sample read cycles are valid, i.e. if the sample read cycle values are
+        equal to the respective run read cycles.
+        Raises:
+            OverrideCyclesError if the reads cycles are not valid.
+        """
+        read1_cycle: str = self.sample_cycles[0]
+        read2_cycle: str = self.sample_cycles[-1]
+        if (
+            read1_cycle == f"Y{self.run_read1_cycles}"
+            and read2_cycle == f"Y{self.run_read2_cycles}"
+        ):
+            return
+        message: str = f"Incorrect read cycles {self.sample_cycles} for sample {self.sample_id}"
+        LOG.error(message)
+        raise OverrideCyclesError(message)
+
+    def _validate_index1_cycles(self) -> None:
+        """
+        Determine if the sample index 1 cycle is valid, i.e., if the number of index characters in
+        the override cycles coincides with the length of the index sequence and if the number of
+        ignored characters in the override cycles matches the difference between the length of the
+        index sequence and the number of run index1 cycles.
+        Raises:
+            OverrideCyclesError if the index 1 cycle is not valid.
+        """
+        index1_cycle: str = self.sample_cycles[1]
+        if (
+            self.run_index1_cycles == len(self.sample[SampleSheetBCLConvertSections.Data.INDEX_1])
+            and index1_cycle == f"I{self.run_index1_cycles}"
+        ):
+            return
+        if self._is_index_cycle_value_following_pattern(
+            pattern=FORWARD_INDEX_CYCLE_PATTERN,
+            index_cycle=index1_cycle,
+            run_cycles=self.run_index1_cycles,
+            index_sequence=self.sample[SampleSheetBCLConvertSections.Data.INDEX_1],
+        ):
+            return
+        message: str = f"Incorrect index1 cycle {index1_cycle} for sample {self.sample_id}"
+        LOG.error(message)
+        raise OverrideCyclesError(message)
+
+    def _validate_index2_cycles(self) -> None:
+        """
+        Determine if the index 2 cycle is valid, i.e., if the number of ignored and index characters
+        correspond to the length of the sample index2 sequence and the number of run index2 cycles,
+        or if the index cycles should be None.
+        Raises:
+            OverrideCyclesError if the index 2 cycle is not valid.
+        """
+        if not self.run_index2_cycles and len(self.sample_cycles) == 3:
+            return
+        index2_cycle: str = self.sample_cycles[2]
+        if (
+            not self.sample[SampleSheetBCLConvertSections.Data.INDEX_2]
+            and index2_cycle == f"N{self.run_index2_cycles}"
```

Comment:
Consider extracting long conditionals to separate utils function with a descriptive name instead

Reply:
I will leave this for another PR